### DISCUSSION TITLE
Issues with the .desktop display in X11 and Wayland

### DISCUSSION THREAD

**Discussion by Bluey26** [lxqt/lxqt#2626] [Q&A] [answered]
Hi, i have noticed that for some reason, the icon display in the screen (the desktop module) is different between labwc(wayland) and X11(openbox).

This means that some items are in different positions in the sessions. I mostly use labwc, but i sometimes enter to X11. If i move  or add 1 .desktop in the X11 session, the order of the icons is distorted in labwc session.

i.e. i have the "Trash" icon in the bottom right corner in labwc, but its icon appears in the upper left corner in X11.

I don't remember if this happened to with wayfire (i have to reinstall it), but i was wondering if this is a known issue. I have not seen any discussions opened in here or in pcmanfm-qt issues zone.

I suspect this issue may have something to do with the "reserved borders" of the screen in both WM's, because the icons that are moved usually are the ones in the bottom and top borders of the screen.

Or maybe its related to the way `~/.config/pcmanfm-qt/lxqt/desktop-items-0.conf` is "read" for each WM?

**Comment by tsujan** [comment, marked as answer]
Do you mean desktop by `.desktop`? The latter is the extension of the mime type "desktop entry" and has no relation to desktop.

When the reserved borders of the desktop are different in X11 and Wayland, the arrangements of desktop items will be different, and logging into each one might affect the arrangement of items in the other.

That's why I use a different profile under Wayland (with labwc, kwin_wayland and wayfire). That's controlled by how pcmanfm-qt's desktop module is started (see `pcmanfm-qt --help`). Different profiles have separate settings inside different folders.

**Reply by Bluey26** [reply]
> Do you mean desktop by `.desktop`? The latter is the extension of the mime type "desktop entry" and has no relation to desktop.

Yes, i mean the desktop, the place in where all the .desktop are(the screen).

I just have checked the borders alignment and they seem to be equal in both sessions(right click in the desktop>Desktop Preferences shows same values for all fields).

Even though the values are the same, i have noticed the margins are not. Its a slightly diference, but it may explain the issue. I have taken the screenshots and made a small .gif for clarity. You can also see that the 3rd .desktop shown is different(without changing the desktop icons positions).

![animation-x11labwc](https://github.com/user-attachments/assets/2893cf4e-b62c-487d-8608-e8a1d5a6d646)

Thats slight difference is "per icon", so in the bottom rows the difference becomes bigger. This makes bottom horizontal line icons to be in the next vertical row.

**Reply by tsujan** [reply]
> I just have checked the borders alignment and they seem to be equal in both sessions

The areas reserved by panels are excluded under X11, but not under Wayland (→ https://github.com/lxqt/pcmanfm-qt/wiki#using-desktop-under-wayland).

**Reply by Bluey26** [reply]
> > I just have checked the borders alignment and they seem to be equal in both sessions
> 
> The areas reserved by panels are excluded under X11, but not under Wayland (→ https://github.com/lxqt/pcmanfm-qt/wiki#using-desktop-under-wayland).

So, if i understand that wiki entry correctly, i should "compensate" that difference between the X11 and the wayland way of Qt geometry?

From your previous comment, i guess this could be done using a pcmanfm-qt profile that does this in wayland and other for x11,right?

**Reply by tsujan** [reply]
If you log between X11 and Wayland like I do, I really recommend a separate profile. I use `pcmanfm-qt --desktop` (without the ending ` --profile=lxqt`), such that the default profile is used on Wayland. Using `pcmanfm-qt --desktop --profile=wayland` is another option.

After doing so, you could manually transfer the settings of the `lxqt` profile to your wayland profile by stopping the desktop module first. Or you might want to set up pcmanfm-qt and its desktop anew under wayland.

**Reply by tsujan** [reply]
> So, if i understand that wiki entry correctly, i should "compensate" that difference between the X11 and the wayland way of Qt geometry?

Yes. Under X11, the desktop items never go behind panels. Under Wayland, you have to set the desktop margins in such a way that items don't go behind panels.

**Reply by Bluey26** [reply]
Excellent, that seems to solve my issue,since the .desktop icons seem to update between each pcmanfm-qt profiles.

What i did was "overriding" `lxqt-desktop.desktop` (inside `~/.config/autostart`) to point to a .sh script, which checks the session (x11 or wayland). The problem i am having with this approach is that "stop" the desktop module does not kill pcmanfm-qt(because its launched by the .sh). I already copied the settings and now everything works as expected besides the "stop" button in session settings.

What would you suggest me to have it working?

the .sh is this:

```
#!/bin/sh
if [ "$XDG_SESSION_TYPE" = "wayland" ];
   then   
       pcmanfm-qt --desktop --profile=wayland
   else
       pcmanfm-qt --desktop --profile=lxqt
fi
```

**Reply by tsujan** [reply]
> since the .desktop icons seem to...

I don't understand the secret of that dot before "desktop" ;) `/usr/share/applications/` if full of those desktop files (entries). Here we're talking about destop **items**.

> ... besides the "stop" button in session settings

You've almost done it:

 * Add `X-LXQt-Module=true` to your desktop entry inside the autostart folder (also, don't forget `OnlyShowIn=LXQt;`). Then, it'll appear among LXQt modules of LXQt Session Settings;
* Also choose a good name for it, like `Name=Desktop for X11 and Wayland`;
* Now, you could easily find it in LXQt Session Settings. Uncheck "Desktop" and check it instead.

**Reply by Bluey26** [reply]
> > since the .desktop icons seem to...
> 
> I don't understand the secret of that dot before "desktop" ;) `/usr/share/applications/` if full of those desktop files (entries). Here we're talking about destop **items**.
> 
> > ... besides the "stop" button in session settings
> 

desktop items, yes. I use .desktop to refer all of those in the screen 😄



> You've almost done it:
> 
>     * Add `X-LXQt-Module=true` to your desktop entry inside the autostart folder (also, don't forget `OnlyShowIn=LXQt;`). Then, it'll appear among LXQt modules of LXQt Session Settings;
> 
>     * Also choose a good name for it, like `Name=Desktop for X11 and Wayland`;
> 
>     * Now, you could easily find it in LXQt Session Settings. Uncheck "Desktop" and check it instead.

Yes, i did the same but i copied the original (`/etc/xdg/autostart/lxqt-desktop.desktop`) inside ~/.config/autostart and modified the `Exec=` line, so this would be "autostarted" each time i log-in in each session.
If i understand it correctly, your approach would need to stop "desktop" every time, and then load "desktop for x11 and wayland" or i am wrong?.

But my problem now is that i need to use `pkill pcmanfm-qt` because `pcmanfm-qt` does not stop if i click the "stop" button in session settings. I was wondering if there was another way to stop the .sh using the Session settings UI(as the other modules).

**Reply by tsujan** [reply]
This is my settings:

I have this `lxqt-desktoploader.desktop` inside the `autostart` folder:

```
[Desktop Entry]
Type=Application
Exec=lxqt-desktoploader
TryExec=lxqt-desktoploader
OnlyShowIn=LXQt;
X-LXQt-Module=true
Name=Desktop for X11 and Wayland
```

`lxqt-desktoploader` is this simple script:

```sh
#!/bin/sh

if [ "$XDG_SESSION_TYPE" = "wayland" ]; then
    exec pcmanfm-qt --desktop
else
    exec pcmanfm-qt --desktop --profile=lxqt
fi
```

It's in my PATH and is made executable. Of course you could use `pcmanfm-qt --desktop --profile=wayland`.

So, I have this:

![desktop](https://github.com/user-attachments/assets/4b956aaa-04cf-4a3f-ac79-2dc1ec182fd0)

As you can guess from the above screenshot, I've done a similar thing for the panel too (but it's a little more complex: four different panel settings for x11, labwc, kwin_wayland and wayfire).

**Reply by Bluey26** [reply]
Oh, i forgot that there were ticks in there to determine the default values for autostart, my bad 😆 . I renamed my autostart file so i have both options, and now the custom one is enabled to autostart instead of "Desktop".

But if you select "Desktop for  X11 and Wayland" and click "stop" button, does something happens? In here it does not work(it does nothing).

So far this is almost complete(though pkill is a simple way to kill pcmanfm-qt). Thank you very much for the help!

**Reply by tsujan** [reply]
> But if you select "Desktop for X11 and Wayland" and click "stop" button, does something happens?

Yes, of course! It's like any other module: it stops and starts pcmanfm-qt's desktop module.

You need to log out and log in, especially because you override `/etc/xdg/autostart/lxqt-desktop.desktop` inside the autostart folder (overriding is a cleaner approach, but I didn't do it for extra reasons).

**Comment by Bluey26** [comment]
For some reason that does not work in here. I even added into the $PATH the bash script, rebooted, but no luck. I will investigate what could be causing this in here later.

**Reply by tsujan** [reply]
FYI
A few minutes ago, I used the overriding approach, and now my LXQt Session Settings is so:

![desktop1](https://github.com/user-attachments/assets/dd25e707-528a-463d-b281-1fefea1a4d7f)

Overriding is cleaner, because you just put the changed desktop entry into `~/.config/autostart` or remove it from there to change the behavior of your module. Logging out and in is necessary for the changes to take effect.

**Comment by Bluey26** [comment]
Btw, do you use the previous (2.0) lxqt-config file? because i do not see the wayland settings in your window.

I ask this because i am using the version with wayland settings, and maybe my issue could be a regression or something, i dont know why the bash script start/stop trigger does not work in here xD

The "desktop" module is the overwritten one. The default ones work as expected.

You can see the desktop items, because pcmanfm-qt is not stopped.

For clarity,check the video:

https://github.com/user-attachments/assets/4ec34dff-61e0-423d-8a5b-483ffa8c6350

**Reply by tsujan** [reply]
> Btw, do you use the previous (2.0) lxqt-config file?

I always use the latest git LXQt — as an LXQt dev, I should.

You don't see the Wayland section in my screenshots only because I don't have `lxqt-wayland-session`  — instead, I use my own Wayland files; my reason is just my experiments with various setups; otherwise, `lxqt-wayland-session` is rock solid and recommended.

Your case is strange to me. If you've put a modified version of `/etc/xdg/autostart/lxqt-desktop.desktop` inside `~/.config/autostart/`, _with the same file name but with contents like in https://github.com/lxqt/lxqt/discussions/2626#discussioncomment-10924668_, then "Desktop" should be replaced by what's in front of `Name=` after you log out and in.

Make sure you've done nothing more than above.

**Reply by tsujan** [reply]
Also make sure that `X-LXQt-Module=true` exists in your desktop file. You once said it's among Autostart entries? **It shouldn't!**

You screencast suggests that pcmanfm-qt may be started twice. Of course, the second try does nothing because there's already an instance running. Make sure that you don't start pcmanfm-qt inside an auto-start script.

**Comment by Bluey26** [comment]
Indeed, its really strange.

Pcmanfm-qt is not present in "Autostart" section under "Session Settings"

This is what i use:
`desktop-profiles` (inside $PATH, variable present in qterminal)

```
#!/bin/sh

if [ "$XDG_SESSION_TYPE" = "wayland" ];
   then
       pcmanfm-qt --desktop --profile=wayland
   else
       pcmanfm-qt --desktop --profile=lxqt
fi
```

In the other hand, i have: 

`~/.config/autostart/lxqt-desktop.desktop`

```
[Desktop Entry]
Type=Application
Exec=desktop-profiles
TryExec=desktop-profiles
OnlyShowIn=LXQt;
X-LXQt-Module=true
Name=Desktop
```

I know the script and the module are working because pcmanfm is launched with the script flags:

`pcmanfm-qt --desktop  --profile=wayland` under labwc, and `pcmanfm-qt --desktop  --profile=lxqt` under openbox.

I also disabled the tick in the "Desktop" module("Basic settings" section), log-out and there was no desktop. Tried to start the module. It loads correctly the desktop module, but stopping it does nothing.

**Reply by tsujan** [reply]
Compare your script with mine: yours lacks `exec`.

**Reply by Bluey26** [reply]
> Compare your script with mine: yours lacks `exec`.

That was the issue, i did not realized that there was an `exec` before each pcmanfm-qt command.

It works now in here, thank you very much 😄

PS: i cannot mark the script comment with "mark as answer" but this is solved.
